File: packages/zig/src/cli/commands/release.zig
//! Release Command — bump version, generate changelog, commit, tag, push.
//!
//! Wraps the `bump` CLI (from zig-bump) and `changelog` CLI (from zig-changelog)
//! into a single `pantry release <type>` command.
//!
//! Flow:
//!   1. Generate changelog via `changelog` (if available)
//!   2. Bump version via `bump <type>` (commits, tags, pushes)
//!   → GitHub Action triggers → creates release → publishes to registry

const std = @import("std");
const io_helper = @import("../../io_helper.zig");
const style = @import("../style.zig");

pub const CommandResult = @import("common.zig").CommandResult;

pub const ReleaseOptions = struct {
    /// Release type: patch, minor, major, premajor, preminor, prepatch, prerelease, or exact version
    release_type: []const u8 = "patch",
    /// Prerelease identifier (alpha, beta, rc)
    preid: ?[]const u8 = null,
    /// Skip confirmation prompts
    yes: bool = false,
    /// Dry run — show what would happen without applying
    dry_run: bool = false,
    /// Skip changelog generation
    no_changelog: bool = false,
    /// Skip git push
    no_push: bool = false,
    /// Custom tag name pattern (default: v{version})
    tag_name: ?[]const u8 = null,
};

/// Find a binary in PATH or pantry's bin directories.
/// Caller owns the returned slice and must free it with `allocator`.
fn findBinary(allocator: std.mem.Allocator, name: []const u8) ?[]const u8 {
    // Check pantry bin dirs first
    const home = io_helper.getenv("HOME") orelse return null;
    const pantry_bin_paths = [_][]const u8{
        "pantry/.bin",
        ".pantry/bin",
    };

    for (pantry_bin_paths) |rel| {
        const full = std.fmt.allocPrint(allocator, "{s}/{s}/{s}", .{ home, rel, name }) catch continue;
        io_helper.accessAbsolute(full, .{}) catch {
            allocator.free(full);
            continue;
        };
        return full;
    }

    // Check CWD pantry/.bin
    const cwd_bin = std.fmt.allocPrint(allocator, "pantry/.bin/{s}", .{name}) catch return null;
    io_helper.accessAbsolute(cwd_bin, .{}) catch {
        allocator.free(cwd_bin);
        // Fall back to PATH lookup via `which`
        return findBinaryInPath(allocator, name);
    };
    return cwd_bin;
}

/// Look up a binary on PATH. Delegates to io_helper.findExecutable, which walks
/// $PATH using libc access(X_OK) — robust across shells (the previous `which`
/// subprocess approach failed because `which` is a shell builtin in some
/// environments, so `bump` couldn't be located even when it was on PATH).
fn findBinaryInPath(allocator: std.mem.Allocator, name: []const u8) ?[]const u8 {
    return (io_helper.findExecutable(allocator, name) catch return null) orelse null;
}

/// Run a git command, returning trimmed stdout (caller owns). Returns null on failure.
fn gitCapture(allocator: std.mem.Allocator, args: []const []const u8) ?[]const u8 {
    const result = io_helper.childRun(allocator, args) catch return null;
    defer allocator.free(result.stderr);
    if (!(result.term == .exited and result.term.exited == 0)) {
        allocator.free(result.stdout);
        return null;
    }
    var end = result.stdout.len;
    while (end > 0 and (result.stdout[end - 1] == '\n' or result.stdout[end - 1] == '\r' or result.stdout[end - 1] == ' ')) {
        end -= 1;
    }
    const out = allocator.dupe(u8, result.stdout[0..end]) catch {
        allocator.free(result.stdout);
        return null;
    };
    allocator.free(result.stdout);
    return out;
}

const SemVer = struct {
    major: u64 = 0,
    minor: u64 = 0,
    patch: u64 = 0,

    /// Parse a semver core (major.minor.patch), ignoring any leading 'v' and
    /// any prerelease/build suffix after the patch number. Returns null if it
    /// doesn't look like a version at all.
    fn parse(s: []const u8) ?SemVer {
        var str = s;
        if (str.len > 0 and (str[0] == 'v' or str[0] == 'V')) str = str[1..];
        var parts: [3]u64 = .{ 0, 0, 0 };
        var idx: usize = 0;
        var num_start: usize = 0;
        var i: usize = 0;
        var saw_digit = false;
        while (i <= str.len) : (i += 1) {
            const at_end = i == str.len;
            const c = if (at_end) '.' else str[i];
            if (c == '.' or c == '-' or c == '+') {
                if (idx < 3) {
                    const seg = str[num_start..i];
                    if (seg.len == 0) return null;
                    parts[idx] = std.fmt.parseInt(u64, seg, 10) catch return null;
                    idx += 1;
                }
                if (c == '-' or c == '+') break;
                num_start = i + 1;
            } else if (c >= '0' and c <= '9') {
                saw_digit = true;
            } else {
                return null;
            }
        }
        if (!saw_digit) return null;
        return .{ .major = parts[0], .minor = parts[1], .patch = parts[2] };
    }

    /// Returns true if `self` is strictly greater than `other`.
    fn greaterThan(self: SemVer, other: SemVer) bool {
        if (self.major != other.major) return self.major > other.major;
        if (self.minor != other.minor) return self.minor > other.minor;
        return self.patch > other.patch;
    }
};

/// Read the `.version = "x.y.z"` value from build.zig.zon in CWD.
/// Caller owns the returned slice.
fn readZonVersion(allocator: std.mem.Allocator) ?[]const u8 {
    const content = io_helper.readFileAlloc(allocator, "build.zig.zon", 1 << 20) catch return null;
    defer allocator.free(content);
    const key = ".version";
    const key_pos = std.mem.indexOf(u8, content, key) orelse return null;
    var i = key_pos + key.len;
    // skip to opening quote
    while (i < content.len and content[i] != '"') : (i += 1) {
        if (content[i] == '\n') return null;
    }
    if (i >= content.len) return null;
    i += 1; // past opening quote
    const start = i;
    while (i < content.len and content[i] != '"') : (i += 1) {}
    if (i >= content.len) return null;
    return allocator.dupe(u8, content[start..i]) catch return null;
}

/// Rewrite the `.version = "..."` value in build.zig.zon in CWD to `new_version`.
/// Returns true on success. Used to sync the zon up to the latest tag before a
/// bump (zig-bump only accepts patch/minor/major, not an explicit version, so we
/// must do the sync ourselves).
fn writeZonVersion(allocator: std.mem.Allocator, new_version: []const u8) bool {
    const content = io_helper.readFileAlloc(allocator, "build.zig.zon", 1 << 20) catch return false;
    defer allocator.free(content);
    const key = ".version";
    const key_pos = std.mem.indexOf(u8, content, key) orelse return false;
    var i = key_pos + key.len;
    while (i < content.len and content[i] != '"') : (i += 1) {
        if (content[i] == '\n') return false;
    }
    if (i >= content.len) return false;
    const val_start = i + 1; // first char inside quotes
    var j = val_start;
    while (j < content.len and content[j] != '"') : (j += 1) {}
    if (j >= content.len) return false;

    const out = std.fmt.allocPrint(allocator, "{s}{s}{s}", .{
        content[0..val_start], new_version, content[j..],
    }) catch return false;
    defer allocator.free(out);

    const file = io_helper.createFile("build.zig.zon", .{}) catch return false;
    defer io_helper.closeFile(file);
    io_helper.writeAllToFile(file, out) catch return false;
    return true;
}

/// Find the highest git tag that parses as a semver (vX.Y.Z). Returns the
/// parsed SemVer, or null if there are no version tags.
fn latestTagVersion(allocator: std.mem.Allocator) ?SemVer {
    const out = gitCapture(allocator, &.{ "git", "tag", "--list", "v*" }) orelse return null;
    defer allocator.free(out);
    var best: ?SemVer = null;
    var it = std.mem.splitScalar(u8, out, '\n');
    while (it.next()) |line| {
        const trimmed = std.mem.trim(u8, line, " \t\r");
        if (trimmed.len == 0) continue;
        const v = SemVer.parse(trimmed) orelse continue;
        if (best == null or v.greaterThan(best.?)) best = v;
    }
    return best;
}

/// Reconcile build.zig.zon version with the latest git tag. zig-bump increments
/// from the zon version and ignores git tags entirely, so if the zon version has
/// drifted *behind* the latest published tag, a `patch` bump would produce a
/// version that's already tagged (a clash) or non-monotonic. To keep releases
/// monotonic across every repo, we sync the zon up to the latest tag's version
/// before bumping, so the increment always lands above the highest existing tag.
fn reconcileVersionDrift(allocator: std.mem.Allocator) void {
    const zon_str = readZonVersion(allocator) orelse return;
    defer allocator.free(zon_str);
    const zon_ver = SemVer.parse(zon_str) orelse return;
    const tag_ver = latestTagVersion(allocator) orelse return;

    if (!tag_ver.greaterThan(zon_ver)) return; // zon is at or ahead of tags — nothing to do

    const synced = std.fmt.allocPrint(allocator, "{d}.{d}.{d}", .{ tag_ver.major, tag_ver.minor, tag_ver.patch }) catch return;
    defer allocator.free(synced);

    style.print(
        "{s}>{s} build.zig.zon version ({s}{s}{s}) is behind latest tag ({s}v{s}{s}); syncing zon to {s}v{s}{s} before bump\n",
        .{
            style.dim,    style.reset, style.yellow,     zon_str, style.reset,
            style.yellow, synced,      style.reset,      style.green_bold,
            synced,       style.reset,
        },
    );

    // zig-bump only accepts patch/minor/major (not an explicit version), and it
    // ignores git tags — so we rewrite the zon ourselves to the latest tag. The
    // subsequent `bump <type>` then increments from there, landing above every
    // existing tag. (Not committed here; the bump commit picks it up.)
    if (!writeZonVersion(allocator, synced)) {
        style.print("  {s}(could not rewrite build.zig.zon — proceeding with bump from {s}){s}\n", .{ style.yellow, zon_str, style.reset });
    }
}

/// Verify the working tree is clean so `bump` only commits the version/changelog
/// changes it makes — not unrelated staged or modified files. Returns true if clean.
fn workingTreeClean(allocator: std.mem.Allocator) bool {
    const out = gitCapture(allocator, &.{ "git", "status", "--porcelain" }) orelse return true;
    defer allocator.free(out);
    // Allow CHANGELOG.md to be dirty — release regenerates it anyway.
    var it = std.mem.splitScalar(u8, out, '\n');
    while (it.next()) |line| {
        const trimmed = std.mem.trim(u8, line, " \t\r");
        if (trimmed.len == 0) continue;
        if (std.mem.endsWith(u8, trimmed, "CHANGELOG.md")) continue;
        return false;
    }
    return true;
}

pub fn releaseCommand(allocator: std.mem.Allocator, options: ReleaseOptions) !CommandResult {
    // Find the bump binary
    const bump_path = findBinary(allocator, "bump") orelse {
        style.print("{s}error:{s} `bump` binary not found.\n", .{ style.red, style.reset });
        style.print("\nInstall it with:\n", .{});
        style.print("  {s}pantry install zig-bump{s}\n\n", .{ style.green_bold, style.reset });
        return .{ .exit_code = 1, .message = null };
    };
    defer allocator.free(bump_path);

    // Guard: refuse to release from a dirty tree (would sweep unrelated changes
    // into the `chore: release` commit). Skipped for dry-run.
    if (!options.dry_run and !workingTreeClean(allocator)) {
        style.print("{s}error:{s} working tree is not clean.\n", .{ style.red, style.reset });
        style.print("  Commit or stash your changes before releasing.\n", .{});
        return .{ .exit_code = 1, .message = null };
    }

    // Reconcile version drift (zon behind latest git tag) so the bump is always
    // monotonic and never clashes with an existing tag. Skipped for dry-run.
    if (!options.dry_run) {
        reconcileVersionDrift(allocator);
    }

    // Step 1: Generate changelog (if changelog binary is available and not skipped)
    if (!options.no_changelog) {
        generateChangelog(allocator);
    }

    // Step 2: Build bump command args (fixed buffer, max 16 args)
    var argv_buf: [16][]const u8 = undefined;
    var argc: usize = 0;

    argv_buf[argc] = bump_path;
    argc += 1;
    argv_buf[argc] = options.release_type;
    argc += 1;
    argv_buf[argc] = "--changelog";
    argc += 1;

    if (options.yes or style.isCI()) {
        argv_buf[argc] = if (style.isCI()) "--ci" else "--yes";
        argc += 1;
    }
    if (options.dry_run) {
        argv_buf[argc] = "--dry-run";
        argc += 1;
    }
    if (options.no_push) {
        argv_buf[argc] = "--no-push";
        argc += 1;
    }
    if (options.preid) |preid| {
        argv_buf[argc] = "--preid";
        argc += 1;
        argv_buf[argc] = preid;
        argc += 1;
    }
    if (options.tag_name) |tag| {
        argv_buf[argc] = "--tag-name";
        argc += 1;
        argv_buf[argc] = tag;
        argc += 1;
    }

    style.print("{s}>{s} bumping version ({s}{s}{s})...\n", .{
        style.dim,        style.reset,
        style.green_bold, options.release_type,
        style.reset,
    });

    // Run bump
    const result = io_helper.childRun(allocator, argv_buf[0..argc]) catch {
        return CommandResult.err(allocator, "Failed to run bump command");
    };
    defer allocator.free(result.stdout);
    defer allocator.free(result.stderr);

    // Print bump output
    if (result.stdout.len > 0) {
        style.print("{s}", .{result.stdout});
    }
    if (result.stderr.len > 0 and !(result.term == .exited and result.term.exited == 0)) {
        style.print("{s}", .{result.stderr});
    }

    if (result.term == .exited and result.term.exited == 0) {
        if (!options.dry_run) {
            style.print("\n{s}{s}{s} release complete\n", .{ style.green, style.check, style.reset });
            if (!options.no_push) {
                style.print("  {s}GitHub Action will create the release and publish to the registry{s}\n", .{ style.dim, style.reset });
            }
        }
        return .{ .exit_code = 0, .message = null };
    }

    return .{ .exit_code = 1, .message = try allocator.dupe(u8, "Release failed") };
}

fn generateChangelog(allocator: std.mem.Allocator) void {
    const changelog_path = findBinary(allocator, "changelog") orelse return;
    defer allocator.free(changelog_path);

    style.print("{s}>{s} generating changelog...\n", .{ style.dim, style.reset });

    const cl_result = io_helper.childRun(allocator, &.{ changelog_path, "-o", "CHANGELOG.md" }) catch {
        style.print("  {s}(changelog generation skipped — command failed){s}\n", .{ style.dim, style.reset });
        return;
    };
    defer allocator.free(cl_result.stdout);
    defer allocator.free(cl_result.stderr);

    if (cl_result.term == .exited and cl_result.term.exited == 0) {
        style.print("  {s}CHANGELOG.md updated{s}\n", .{ style.green, style.reset });
    } else {
        style.print("  {s}(changelog generation skipped){s}\n", .{ style.dim, style.reset });
    }
}

File: packages/ts-pantry/src/release.ts
import { spawnSync } from 'node:child_process'
import fs from 'node:fs'
import os from 'node:os'
import path from 'node:path'
import process from 'node:process'

export type ReleaseType = 'patch' | 'minor' | 'major'

export interface ReleaseOptions {
  /** Release type: patch | minor | major */
  type: ReleaseType
  /** Non-interactive — skip confirmation prompts */
  yes?: boolean
  /** Preview the actions without applying any change */
  dryRun?: boolean
  /** Skip changelog generation */
  noChangelog?: boolean
  /** Skip the final `git push` (commit + tag stay local) */
  noPush?: boolean
  /** Skip the publish step entirely */
  noPublish?: boolean
  /** Working directory of the zig package (default: process.cwd()) */
  cwd?: string
}

export interface ReleaseResult {
  exitCode: number
}

const RED = '\x1B[31m'
const GREEN = '\x1B[32m'
const YELLOW = '\x1B[33m'
const DIM = '\x1B[2m'
const BOLD = '\x1B[1m'
const RESET = '\x1B[0m'

function info(msg: string): void {
  console.log(`${DIM}>${RESET} ${msg}`)
}

function ok(msg: string): void {
  console.log(`${GREEN}✓${RESET} ${msg}`)
}

function warn(msg: string): void {
  console.log(`${YELLOW}!${RESET} ${msg}`)
}

function fail(msg: string): void {
  console.error(`${RED}error:${RESET} ${msg}`)
}

interface RunResult {
  status: number
  stdout: string
  stderr: string
}

function run(cmd: string, args: string[], cwd: string): RunResult {
  const res = spawnSync(cmd, args, { cwd, encoding: 'utf8' })
  return {
    status: res.status ?? (res.error ? 1 : 0),
    stdout: res.stdout ?? '',
    stderr: res.stderr ?? (res.error ? String(res.error.message) : ''),
  }
}

/**
 * Resolve a tool binary (`bump`, `changelog`) by checking, in order:
 *   1. PATH (via `which`)
 *   2. pantry bin dirs (~/.local/bin, ~/pantry/.bin, ~/.pantry/bin)
 *   3. the user's local zig source checkouts (~/Code/Libraries/<repo>/zig-out/bin/<name>)
 */
function findTool(name: string, repoHint: string): string | null {
  const which = spawnSync('which', [name], { encoding: 'utf8' })
  if ((which.status ?? 1) === 0 && which.stdout.trim().length > 0)
    return which.stdout.trim()

  const home = os.homedir()
  const candidates = [
    path.join(home, '.local', 'bin', name),
    path.join(home, 'pantry', '.bin', name),
    path.join(home, '.pantry', 'bin', name),
    path.join(home, 'Code', 'Libraries', repoHint, 'zig-out', 'bin', name),
  ]
  for (const c of candidates) {
    if (fs.existsSync(c))
      return c
  }
  return null
}

/** Read the `.version = "x.y.z"` field from a build.zig.zon file. */
function readZonVersion(zonPath: string): string | null {
  const src = fs.readFileSync(zonPath, 'utf8')
  const m = src.match(/\.version\s*=\s*"([^"]+)"/)
  return m ? m[1] : null
}

/** Sync the `version` field in package.json (if present) to `version`. */
function syncPackageJsonVersion(cwd: string, version: string): boolean {
  const pkgPath = path.join(cwd, 'package.json')
  if (!fs.existsSync(pkgPath))
    return false
  const raw = fs.readFileSync(pkgPath, 'utf8')
  // Preserve formatting by doing a targeted replace of the top-level version field.
  const updated = raw.replace(/("version"\s*:\s*)"[^"]*"/, `$1"${version}"`)
  if (updated === raw)
    return false
  fs.writeFileSync(pkgPath, updated)
  return true
}

/**
 * Run a full release for a Zig package living in `cwd`:
 *   changelog → bump version (file only) → sync package.json → commit → tag → push → publish.
 */
export async function runRelease(options: ReleaseOptions): Promise<ReleaseResult> {
  const cwd = options.cwd ?? process.cwd()
  const zonPath = path.join(cwd, 'build.zig.zon')

  // 1. Detect a zig package.
  if (!fs.existsSync(zonPath)) {
    fail(`no build.zig.zon found in ${cwd}`)
    console.error(`  ${DIM}\`pantry release\` must be run from the root of a Zig package.${RESET}`)
    return { exitCode: 1 }
  }

  // 2. Must be inside a git repo.
  const insideGit = run('git', ['rev-parse', '--is-inside-work-tree'], cwd)
  if (insideGit.status !== 0 || insideGit.stdout.trim() !== 'true') {
    fail('not inside a git repository')
    return { exitCode: 1 }
  }

  // 3. Working tree must be clean (so the release commit is exactly the bump).
  const dirty = run('git', ['status', '--porcelain'], cwd)
  if (dirty.status === 0 && dirty.stdout.trim().length > 0 && !options.dryRun) {
    fail('working tree is not clean — commit or stash your changes first.')
    console.error(dirty.stdout.trimEnd())
    return { exitCode: 1 }
  }

  // 4. Verify there is an upstream to push to (unless pushing is skipped).
  let hasUpstream = false
  if (!options.noPush) {
    const upstream = run('git', ['rev-parse', '--abbrev-ref', '--symbolic-full-name', '@{u}'], cwd)
    hasUpstream = upstream.status === 0 && upstream.stdout.trim().length > 0
    if (!hasUpstream)
      warn('no upstream branch configured — will commit and tag locally (skipping push).')
  }

  const currentVersion = readZonVersion(zonPath)
  if (!currentVersion) {
    fail(`could not read .version from ${zonPath}`)
    return { exitCode: 1 }
  }

  // 5. Locate the bump binary (from zig-bump).
  const bump = findTool('bump', 'zig-bump')
  if (!bump) {
    fail('`bump` binary not found (from zig-bump).')
    console.error(`  ${DIM}Build it: cd ~/Code/Libraries/zig-bump && zig build, or install it onto your PATH.${RESET}`)
    return { exitCode: 1 }
  }

  // 6. Generate / update the changelog (best-effort) from conventional commits.
  if (!options.noChangelog) {
    const changelog = findTool('changelog', 'zig-changelog')
    if (changelog) {
      info('generating changelog from conventional commits...')
      if (options.dryRun) {
        console.log(`  ${DIM}[dry-run] would run: changelog -o CHANGELOG.md${RESET}`)
      }
      else {
        const cl = run(changelog, ['-o', 'CHANGELOG.md'], cwd)
        if (cl.status === 0)
          ok('CHANGELOG.md updated')
        else
          warn(`changelog generation skipped${cl.stderr ? `: ${cl.stderr.trim()}` : ''}`)
      }
    }
    else {
      warn('`changelog` binary not found (from zig-changelog) — skipping changelog.')
    }
  }

  // 7. Bump the version in build.zig.zon (file only; we own the git steps).
  info(`bumping version (${BOLD}${options.type}${RESET}) from ${currentVersion}...`)
  if (options.dryRun) {
    const dry = run(bump, [options.type, '--dry-run', '--yes'], cwd)
    if (dry.stdout.trim())
      console.log(dry.stdout.trimEnd())
    console.log(`  ${DIM}[dry-run] no files changed, no commit/tag/push, no publish.${RESET}`)
    return { exitCode: 0 }
  }

  const bumpRes = run(bump, [options.type, '--no-commit', '--no-tag', '--no-push', '--yes'], cwd)
  if (bumpRes.status !== 0) {
    fail('version bump failed')
    if (bumpRes.stdout.trim())
      console.error(bumpRes.stdout.trimEnd())
    if (bumpRes.stderr.trim())
      console.error(bumpRes.stderr.trimEnd())
    return { exitCode: 1 }
  }

  const newVersion = readZonVersion(zonPath)
  if (!newVersion || newVersion === currentVersion) {
    fail('version did not change after bump')
    return { exitCode: 1 }
  }
  ok(`build.zig.zon ${currentVersion} → ${newVersion}`)

  const tag = `v${newVersion}`

  // Refuse to clobber an existing tag.
  const tagExists = run('git', ['rev-parse', '-q', '--verify', `refs/tags/${tag}`], cwd)
  if (tagExists.status === 0) {
    fail(`tag ${tag} already exists — bump to a version that has no tag yet.`)
    // Roll back the file change so the tree is left clean.
    run('git', ['checkout', '--', 'build.zig.zon'], cwd)
    return { exitCode: 1 }
  }

  // 8. Sync package.json version (if present).
  if (syncPackageJsonVersion(cwd, newVersion))
    ok(`package.json version → ${newVersion}`)

  // 9. Stage + commit the release.
  run('git', ['add', 'build.zig.zon'], cwd)
  if (fs.existsSync(path.join(cwd, 'package.json')))
    run('git', ['add', 'package.json'], cwd)
  if (fs.existsSync(path.join(cwd, 'CHANGELOG.md')))
    run('git', ['add', 'CHANGELOG.md'], cwd)

  const commit = run('git', ['commit', '-m', `chore: release ${tag}`], cwd)
  if (commit.status !== 0) {
    fail('git commit failed')
    if (commit.stderr.trim())
      console.error(commit.stderr.trimEnd())
    return { exitCode: 1 }
  }
  ok(`committed release ${tag}`)

  // 10. Create the annotated tag.
  const tagRes = run('git', ['tag', '-a', tag, '-m', `Release ${tag}`], cwd)
  if (tagRes.status !== 0) {
    fail('git tag failed')
    if (tagRes.stderr.trim())
      console.error(tagRes.stderr.trimEnd())
    return { exitCode: 1 }
  }
  ok(`tagged ${tag}`)

  // 11. Push the commit and the tag.
  if (!options.noPush && hasUpstream) {
    info('pushing commit and tag...')
    const pushCommit = run('git', ['push'], cwd)
    if (pushCommit.status !== 0) {
      fail('git push failed')
      if (pushCommit.stderr.trim())
        console.error(pushCommit.stderr.trimEnd())
      return { exitCode: 1 }
    }
    const pushTag = run('git', ['push', 'origin', tag], cwd)
    if (pushTag.status !== 0) {
      fail(`pushing tag ${tag} failed`)
      if (pushTag.stderr.trim())
        console.error(pushTag.stderr.trimEnd())
      return { exitCode: 1 }
    }
    ok(`pushed commit and tag ${tag}`)
  }
  else {
    warn('skipped push (commit and tag are local only).')
  }

  // 12. Publish to the pantry registry (best-effort / skippable).
  if (!options.noPublish)
    publish(cwd)

  console.log(`\n${GREEN}${BOLD}✓ released ${tag}${RESET}`)
  return { exitCode: 0 }
}

/**
 * Best-effort publish to the pantry registry. The `pantry publish` flow for Zig
 * packages lives in the Zig CLI (`packages/zig`, `pantry publish`). We shell out
 * to whatever `pantry` is on PATH; if it lacks a working `publish` for zig yet,
 * we report it clearly and let the GitHub Action (triggered by the pushed tag)
 * handle publishing instead.
 */
function publish(cwd: string): void {
  const pantry = findTool('pantry', 'pantry')
  if (!pantry) {
    warn('`pantry` binary not found — skipping publish (the pushed tag\'s CI will publish).')
    return
  }
  // Probe: does this pantry build expose a `publish` command?
  const help = run(pantry, ['publish', '--help'], cwd)
  if (help.status !== 0) {
    warn('`pantry publish` is not available in the installed pantry yet — skipping.')
    console.log(`  ${DIM}The pushed tag will trigger CI to publish to the registry.${RESET}`)
    return
  }
  info('publishing to pantry registry...')
  const pub = run(pantry, ['publish'], cwd)
  if (pub.stdout.trim())
    console.log(pub.stdout.trimEnd())
  if (pub.status === 0) {
    ok('published to pantry registry')
  }
  else {
    warn('publish step failed — the pushed tag\'s CI can still publish.')
    if (pub.stderr.trim())
      console.log(`  ${DIM}${pub.stderr.trim()}${RESET}`)
  }
}
